File: src/main/java/com/dude/dms/ui/components/cards/PlainTextRuleCard.java
package com.dude.dms.ui.components.cards;

import com.dude.dms.backend.brain.parsing.PlainTextRuleValidator;
import com.dude.dms.backend.data.rules.PlainTextRule;
import com.dude.dms.backend.service.TagService;
import com.dude.dms.ui.builder.BuilderFactory;
import com.dude.dms.ui.components.tags.TagContainer;
import com.github.appreciated.card.RippleClickableCard;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;

public class PlainTextRuleCard extends RippleClickableCard {

    public PlainTextRuleCard(BuilderFactory builderFactory, PlainTextRule rule, PlainTextRuleValidator plainTextRuleValidator, TagService tagService) {
        setWidthFull();

        TagContainer tagContainer = new TagContainer(tagService.findByPlainTextRule(rule));
        tagContainer.setPadding(true);

        Button runButton = new Button("Run", VaadinIcon.PLAY.create(), e -> builderFactory.rules().ruleRunnerDialog(plainTextRuleValidator.runRuleForAll(rule)).build().open());
        runButton.addThemeVariants(ButtonVariant.LUMO_SUCCESS);

        HorizontalLayout wrapper = new HorizontalLayout(runButton, new Label(rule.getText()), tagContainer);
        wrapper.setWidthFull();
        wrapper.setMinHeight("10vh");
        wrapper.setAlignItems(Alignment.CENTER);
        wrapper.setPadding(true);
        add(wrapper);
    }
}

File: src/main/kotlin/com/dude/dms/ui/components/cards/PlainTextRuleCard.kt
package com.dude.dms.ui.components.cards

import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.filter.PlainTextRule
import com.dude.dms.brain.events.EventType
import com.dude.dms.brain.t
import com.dude.dms.utils.*
import com.dude.dms.ui.components.tags.TagLayout
import com.github.appreciated.card.RippleClickableCard
import com.github.mvysny.karibudsl.v10.button
import com.github.mvysny.karibudsl.v10.horizontalLayout
import com.github.mvysny.karibudsl.v10.label
import com.github.mvysny.karibudsl.v10.onLeftClick
import com.vaadin.flow.component.button.ButtonVariant
import com.vaadin.flow.component.html.Label
import com.vaadin.flow.component.icon.VaadinIcon
import com.vaadin.flow.component.orderedlayout.FlexComponent

class PlainTextRuleCard(rule: PlainTextRule) : RippleClickableCard() {

    private lateinit var label: Label
    private lateinit var tagContainer: TagLayout

    init {
        eventManager.register(this, Tag::class, EventType.UPDATE, EventType.DELETE) { fill(rule) }
        setWidthFull()
        addClickListener { plaintextRuleEditDialog(rule).open() }

        horizontalLayout(isPadding = true) {
            setWidthFull()
            alignItems = FlexComponent.Alignment.CENTER

            button(t("run"), VaadinIcon.PLAY.create()) {
                onLeftClick {
                    this@horizontalLayout.ruleRunnerDialog(plainTextRuleValidator.runRuleForAll(rule)).open()
                }
                addThemeVariants(ButtonVariant.LUMO_SUCCESS)
            }
            label = label(rule.text)
            tagContainer = tagLayout(rule.tags.toMutableSet())
        }
    }

    fun fill(rule: PlainTextRule) {
        tagContainer.tags = rule.tags.toMutableSet()
        label.text = rule.text
    }
}